r/R/dataset.R

```diff
@@ -141,6 +141,42 @@ Dataset$create <- function(children, schema) {
 #' @export
 names.Dataset <- function(x) names(x$schema)
 
+#' @export
+dim.Dataset <- function(x) {
+
+  if (!inherits(x$format, "ParquetFileFormat")) {
+    stop(
+      "dim() is only currently implemented for Datasets that use parquet files.",
+      "Call collect() first to pull data into R.",
+      call. = FALSE
+    )
+  }
+
+  rows <- as.integer(sum(purrr::map_dbl(x$files, ~ParquetFileReader$create(.x)$num_rows)))
```

Comment:
```suggestion
  rows <- sum(map_int(x$files, ~ParquetFileReader$create(.x)$num_rows))
```

Reply:
I now remember that I tried this earlier and since `ParquetFileReader$create(.x)$num_rows` returns a number `map_int` fails. If I coerced [here](https://github.com/boshek/arrow/blob/01432c8077cb7286a3c5db6d3a8a679d0b88e5d4/r/R/parquet.R#L448) do you think that would be better?

Reply:
I ended coercing there. That at least passes the tests.

Reply:
That's probably fine. IIUC Rcpp translates the `int64_t` that the C++ function returns into an R `double` because `integer` is only 32-bits so there's a chance it would overflow. I think that's highly unlikely to happen, so this coercion is safe in practice.